pandas/core/nanops.py

```diff
@@ -254,9 +254,16 @@ def nansum(values, axis=None, skipna=True):
 @bottleneck_switch()
 def nanmean(values, axis=None, skipna=True):
     values, mask, dtype, dtype_max = _get_values(values, skipna, 0)
-    the_sum = _ensure_numeric(values.sum(axis, dtype=dtype_max))
     count = _get_counts(mask, axis)
 
+    dtype_sum = dtype_max
+    if is_integer_dtype(dtype):
+        dtype_sum = np.float64
+    elif is_float_dtype(dtype):
+        dtype_sum = dtype
+        count = dtype.type(count)
```

Comment:
hmm, seems we are casting that, don't really remember why. You can try changing ``get_count`` to simply return an ``int64`` which I think will always be correct.